### PR TITLE
feat(transaction/validate): advance transaction validation when producing blocks 

### LINKED ISSUE #5564: Advance transaction validation when producing blocks
# Rationale

Why should this feature exist?
Transaction execution requires some resources to be requested, and this process can be time-consuming, so transaction validation should be done in advance to avoid unnecessary waste of resources.

## Current Transaction Validation
Tapos
TooBig
Expiration
ContractSize
Signature
Duplication

## Current Block Production Flow
image

# Implementation

Do you have ideas regarding the implementation of this feature?
 Transactions are validated first and executed after.
## New Block Production Flow
image

Are you willing to implement this feature?
Yes

### PR DESCRIPTION
**Pull request by halibobo1205**
close #5564.

### CODE REVIEW

File: framework/src/main/java/org/tron/common/validator/BigTransactionValidator.java [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+import org.tron.core.capsule.TransactionCapsule;
+
+@Component("bigTransactionValidator")
+public class BigTransactionValidator extends AbstractTransactionValidator {
+
+  @Override
+  protected Pair<GrpcAPI.Return.response_code, String> doValidate(TransactionCapsule trx) {
+    if (trx.getSerializedSize() > Constant.TRANSACTION_MAX_BYTE_SIZE) {
+      return buildResponse(GrpcAPI.Return.response_code.TOO_BIG_TRANSACTION_ERROR,
+          "too big transaction, the size is %d bytes", trx.getSerializedSize());
```
[the remark this answers is not in the document]
Comment (halibobo1205) [reply framework/src/main/java/org/tron/common/validator/BigTransactionValidator.java:16]: 1.  check size don't need to ask for new arrays.
image
image

2.  accepted!

File: framework/src/main/java/org/tron/common/validator/DupTransactionValidator.java [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+  @Autowired
+  private TransactionStore transactionStore;
+
+  @Autowired
+  private TransactionCache transactionCache;
+
+  @Override
+  protected Pair<GrpcAPI.Return.response_code, String> doValidate(TransactionCapsule trx) {
+    byte[] transactionId = trx.getTransactionId().getBytes();
+    if (transactionCache.has(transactionId) && transactionStore.has(transactionId)) {
```
[the remark this answers is not in the document]
Comment (halibobo1205) [reply framework/src/main/java/org/tron/common/validator/DupTransactionValidator.java:23]: transaction cache uses Bloom filters, which have a false positive rate when found.

[the remark this answers is not in the document]
Comment (halibobo1205) [reply framework/src/main/java/org/tron/common/validator/DupTransactionValidator.java:23]: https://github.com/tronprotocol/java-tron/pull/5582/files#diff-07030dab54a909bcef6a29ff60e124374c426fe8c119e80f265e68d92fc9cc09L818

File: framework/src/main/java/org/tron/common/validator/BigTransactionValidator.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (halibobo1205) [reply framework/src/main/java/org/tron/common/validator/BigTransactionValidator.java]: done.

File: framework/src/main/java/org/tron/common/validator/SignatureValidator.java [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+  @Autowired
+  private DynamicPropertiesStore dynamicPropertiesStore;
+
+  @Override
+  protected Pair<GrpcAPI.Return.response_code, String> doValidate(TransactionCapsule trx) {
+    try {
+      trx.validateSignature(accountStore, dynamicPropertiesStore);
+      return SUCCESS;
+    } catch (ValidateSignatureException e) {
+      return buildResponse(GrpcAPI.Return.response_code.SIGERROR, e.getMessage());
```
[the remark this answers is not in the document]
Comment (halibobo1205) [reply framework/src/main/java/org/tron/common/validator/SignatureValidator.java:26]: `trx.validateSignature(AccountStore accountStore,
      DynamicPropertiesStore dynamicPropertiesStore)` never returns false,  the following are not valid.

image

File: framework/src/main/java/org/tron/core/db/Manager.java [hunk 4]
```diff
@@ -1597,6 +1534,10 @@ public BlockCapsule generateBlock(Miner miner, long blockTime, long timeout) {
         postponedTrxCount++;
         continue; // try pack more small trx
       }
+      // check transaction
+      if (!transactionValidator.silentValidate(trx)) {
```
Comment (lurais) [inline framework/src/main/java/org/tron/core/db/Manager.java:1538]: Actually, some checks may be skipped if that has been checked when it was pushed into pending queue.

Comment (halibobo1205) [reply framework/src/main/java/org/tron/core/db/Manager.java:1538]: Not necessary for now.

File: framework/src/main/java/org/tron/common/validator/AbstractTransactionValidator.java [hunk 5]
```diff
@@ -0,0 +1,34 @@
+package org.tron.common.validator;
+
+import org.apache.commons.lang3.tuple.Pair;
+import org.tron.api.GrpcAPI.Return.response_code;
+import org.tron.core.capsule.TransactionCapsule;
+
+public abstract class AbstractTransactionValidator implements Validator<
+    Pair<response_code, String>, TransactionCapsule> {
```
[the remark this answers is not in the document]
Comment (halibobo1205) [reply framework/src/main/java/org/tron/common/validator/AbstractTransactionValidator.java:8]: `response_code `  is generated by proto.

**Review by jwrct — APPROVED** [review]

**Comment by halibobo1205** [comment]
There is no significant performance improvement, closed for now.

**Pull request closed without merging**